### ISSUE TITLE
No-fallthrough doesn't work

### ISSUE THREAD

**Issue by aylusltd** [eslint/eslint#4950]
The following code trips up no-fallthrough for a false positive:

```javascript
    function test(thing) {
        switch(thing) {
            case 1:
                if (true) {
                    return true;
                } else {
                    return false;
                }
        }
    }
```

Obviously it can't fall through since either branch returns.

**Comment by eslintbot** [comment]
@aylusltd Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by aylusltd** [comment]
1. v1.6.0 is what I'm running
2. see above for code, config below:
```yaml
    no-fallthrough: 1
```
3. `Expected a "break" statement before "case" (no-fallthrough)` (It's in sublime text, so not c/p)
4. As noted above, should not fail.

**Comment by ilyavolodin** [comment]
v1.6.0 is a very old version. This rule has been re-written since then to use code-path analysis. Could you try latest version to see if the issue still occurs?

**Comment by aylusltd** [comment]
Updating did not fix it.

v 1.10.3 running now

**Comment by gyandeeps** [comment]
Cannot reproduce this in [online demo](http://eslint.org/demo/) using `1.10.3`.

**Comment by nzakas** [comment]
@aylusltd can you share the actual code and complete output? I'm guessing the code you have above is meant to be a simplified repro case, but we can't reproduce with it.

**Comment by aylusltd** [comment]
```javascript

var w=[];

function makeMountain(y, x) {
        if (w[y][x] === 'mountain') {
            return w;
        }

        w[y][x] = 'mountain';
        var move = Math.floor(Math.random() * 8);
        switch (move) {
            case 0:
                // up left
                if (y && x) {
                    return makeMountain(y - 1, x - 1);
                } else {
                    return w;
                }
            case 1:
                // up
                if (y) {
                    return makeMountain(y - 1, x);
                } else {
                    return w;
                }
            case 2:
                // up right
                if (y && x < w[y].length - 1) {
                    return makeMountain(y - 1, x + 1);
                } else {
                    return w;
                }
            case 3:
                // right
                if (x < w[y].length - 1) {
                    return makeMountain(y, x + 1);
                } else {
                    return w;
                }
            case 4:
                // down right
                if (y < w.length - 1 && x < w[y].length - 1) {
                    return makeMountain(y + 1, x + 1);
                } else {
                    return w;
                }
            case 5:
                // down
                if (y < w.length - 1) {
                    return makeMountain(y+1,x);
                } else {
                    return w;
                }
            case 6:
                // down left
                if (x && y < w.length - 1) {
                    return makeMountain(y + 1, x - 1);
                } else {
                    return w;
                }
            case 7:
                // left
                if (x) {
                    return makeMountain(y, x - 1);
                } else {
                    return w;
                }
            default:
                return w;
        }
    }
makeMountain(1,2);

```

**Comment by gyandeeps** [comment]
Yes this code fails with `1.10.3` but passes with current `master`.
@aylusltd Can you try it with current `master` just to confirm?

**Comment by aylusltd** [comment]
Will do.

**Comment by platinumazure** [comment]
@aylusltd Did you ever note which lines in your example are being flagged by ESLint? Or did I miss that?

**Comment by aylusltd** [comment]
@platinumazure 

18:13 - Expected a "break" statement before "case". (no-fallthrough)
25:13 - Expected a "break" statement before "case". (no-fallthrough)
32:13 - Expected a "break" statement before "case". (no-fallthrough)
39:13 - Expected a "break" statement before "case". (no-fallthrough)
46:13 - Expected a "break" statement before "case". (no-fallthrough)
53:13 - Expected a "break" statement before "case". (no-fallthrough)
60:13 - Expected a "break" statement before "case". (no-fallthrough)
67:13 - Expected a "break" statement before "default". (no-fallthrough)

**Comment by aylusltd** [comment]
@gyandeeps is there a way to make sure no-fallthrough is set to 1 or 2 from the CLI? I don't see the error from the latest master, but it's not using my .eslintrc

**Comment by mysticatea** [comment]
Yes, ESLint v1 is not checking inside of `if` block.
I think this has been fixed by #4251.

**Comment by mysticatea** [comment]
```
> eslint -v
v1.10.3
> eslint test.js --no-ignore --no-eslintrc --rule no-fallthrough:2

C:\Users\t-nagashima.AD\Documents\GitHub\eslint\test.js
  18:13  error  Expected a "break" statement before "case"     no-fallthrough
  25:13  error  Expected a "break" statement before "case"     no-fallthrough
  32:13  error  Expected a "break" statement before "case"     no-fallthrough
  39:13  error  Expected a "break" statement before "case"     no-fallthrough
  46:13  error  Expected a "break" statement before "case"     no-fallthrough
  53:13  error  Expected a "break" statement before "case"     no-fallthrough
  60:13  error  Expected a "break" statement before "case"     no-fallthrough
  67:13  error  Expected a "break" statement before "default"  no-fallthrough

✖ 8 problems (8 errors, 0 warnings)

> npm i eslint@next
....
> eslint -v
v2.0.0-beta.1
> eslint test.js --no-ignore --no-eslintrc --rule no-fallthrough:2
>
```
